### PR TITLE
add support for i586-unknown-linux-gnu

### LINKED ISSUE #156: Support i586-unknown-linux-gnu

### PR DESCRIPTION
**Pull request by gnzlbg**
Closes #156 .

### CODE REVIEW

**Comment by gnzlbg** [comment]
I don't have a linux machine, is there a way to get the tests to run? Or do I have to test it manually (e.g. by forking cargo and trying to compile it with cross for i586 ?)

File: docker/i586-unknown-linux-gnu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:17.04
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+    ca-certificates \
+    cmake \
+    gcc \
+    libc6-dev \
+    make \
+    file \
```
Comment (malbarbo) [inline docker/i586-unknown-linux-gnu/Dockerfile]: Is file dependency required?

File: docker/i586-unknown-linux-gnu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:17.04
```
Comment (malbarbo) [inline docker/i586-unknown-linux-gnu/Dockerfile]: This file is based on i686-unknown-linux-gnu, right? If the ubuntu version is kept in 12.04, the docker layers can be shared between i686-unknown-linux-gnu and i586-unknown-linux-gnu.

File: README.md [hunk 3]
```diff
@@ -198,6 +198,7 @@ worst, "hang" (never terminate).
 | `armv7-unknown-linux-gnueabihf`      | 2.15   | 4.6.2   | 1.0.2m  | ✓   | 2.8.0 |   ✓    |
 | `armv7-unknown-linux-musleabihf`     | 1.1.15 | 5.3.1   | N/A     |     | 2.8.0 |   ✓    |
 | `asmjs-unknown-emscripten` [4]       | 1.1.15 | 1.37.13 | N/A     | ✓   | N/A   |   ✓    |
+| `i586-unknown-linux-gnu`             | 2.15   | 4.6.2   | 1.0.2m  | ✓   | N/A   |   ✓    |
```
Comment (japaric) [inline README.md]: the glibc version is wrong. Ubuntu 17.04 ships with glibc 2.24.

File: docker/i586-unknown-linux-gnu/Dockerfile (hunk 2 above)
Comment (gnzlbg) [reply docker/i586-unknown-linux-gnu/Dockerfile]: I don't really know my way around docker. Should I downgrade the version in this file, upgrade the version in the `i686` docker file, or do I need to do something else to reuse the docker layers? 

(FWIW in the `stdsimd` crate we use 17.04 for everything and I am trying to replace our testing infrastructure there with `cross` now that you are adding support here for qemu-system).

File: docker/i586-unknown-linux-gnu/Dockerfile (hunk 1 above)
Comment (gnzlbg) [reply docker/i586-unknown-linux-gnu/Dockerfile]: No, this is a c&p-error from the `i586` file in the `stdsimd` crate.

File: docker/i586-unknown-linux-gnu/Dockerfile (hunk 2 above)
Comment (japaric) [inline docker/i586-unknown-linux-gnu/Dockerfile]: This should be as low as possible, i.e. 12.04 as the i686 target, to make the binaries produced with Cross compatible with legacy distributions like centos.

**Comment by malbarbo** [comment]
I don't known how to run the test in a non Linux machine... Maybe you can install Linux in a virtual machine?

Instructions to run the tests in a Linux machine:

First you need to build the docker image:

```
./build-docker-image.sh i586-unknown-linux-gnu
```

Then you can run the tests:

```
TRAVIS_OS_NAME=linux TARGET=i586-unknown-linux-gnu CPP=1 DYLIB=1 STD=1 OPENSSL=0.5.5 RUN=1 bash ci/script.sh
```

**Comment by japaric** [comment]
> I don't have a linux machine, is there a way to get the tests to run?

I can throw this into homu as a try build but test times will be in the order of hours due to the large number of travis builders. If you temporarily comment all the other travis builds I can start a try build.

File: docker/i586-unknown-linux-gnu/Dockerfile (hunk 2 above)
Comment (gnzlbg) [reply docker/i586-unknown-linux-gnu/Dockerfile]: Fixed. This should make the version in the readme file correct right?

Comment (gnzlbg) [reply docker/i586-unknown-linux-gnu/Dockerfile]: Per @japaric 's comment below I've downgraded this to 12.04.

**Comment by gnzlbg** [comment]
@japaric I have another PR with an ARM build, maybe after that one passes review I can merge both PRs and then we can do a homu build of both together?

File: README.md (hunk 3 above)
Comment (gnzlbg) [reply README.md]: I've downgraded the ubuntu version to 12.04, so that this should now be correct.

**Comment by japaric** [comment]
@gnzlbg Fine by me.

**Comment by gnzlbg** [comment]
@japaric I've merged both PR here and disabled the other builds so that homu can be run.

**Comment by gnzlbg** [comment]
@japaric is there a way to force qemu-arm to emulate an armv6 processor ?

**Comment by gnzlbg** [comment]
@japaric I've removed the arm target since I don't really know how to do that without a machine where I can test it, the i586 parts should be good to go if you want to try with homu.

**Comment by japaric** [comment]
> is there a way to force qemu-arm to emulate an armv6 processor ?

I think I tried setting a specific CPU before using QEMU_CPU but I think I didn't manage to get a SIGILL when supossedly emulating ARMv6 and running a binary that contained ARMv7-only instructions.

> I don't really know how to do that without a machine where I can test it

You can probably copy paste some of the Docker build stuff from rust-lang/rust. There they use crosstool-ng to build an ARMv6 toolchain.

@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 81f03cd510640c341df0005cc7f21b86cadae43a with merge 7e225bd6303bde0bc2f99e44270a47aa1763b883...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/302095949?utm_source=github_status&utm_medium=notification)

**Comment by gnzlbg** [comment]
It fails to link `cargo`: https://travis-ci.org/japaric/cross/builds/302095949#L8670

**Comment by japaric** [comment]
That looks slightly familiar. IIRC that error is caused because you used a relatively new binutils / gcc when compiling the std facade but used an old binutils (ld) when linking a Rust binary against the facade. This has happened before with 32-bit targets and it was fixed upstream, perhaps this target was not fixed? I believe the solution was to downgrade the binutils version used in the docker image (in rust-lang/rust) where std is built.

**Comment by gnzlbg** [comment]
@japaric you mean this here, and the comment above?

https://travis-ci.org/japaric/cross/builds/302095949#L8670

It seems like that is done for i686 but not for i586.

**Comment by japaric** [comment]
I was referring to this linker error:

>   = note: /usr/bin/ld: /rust/lib/rustlib/i586-unknown-linux-gnu/lib/liballoc_jemalloc-208c05c6a7233cbb.rlib(jemalloc.pic.o): unrecognized relocation (0x2b) in section `.text.malloc_conf_init'

Specifically the "unrecognized relocation" you can probably find issues with that phrase in rust-lang/rust.

**Comment by gnzlbg** [comment]
Ok I c&p the wrong link, this is the right one: https://github.com/rust-lang/rust/blob/master/src/ci/docker/dist-i586-gnu-i686-musl/Dockerfile#L36

**Comment by japaric** [comment]
Yes, that's the issue. The problem is that std is built in Ubuntu 16.04 and we are using a linker from 12.04.

> ENV CFLAGS_i686_unknown_linux_musl=-Wa,-mrelax-relocations=no

This flag should also be applied to i586-gnu otherwise only i686-musl would be "supercompatible". This has to be fixed upstream.

**Comment by gnzlbg** [comment]
https://github.com/rust-lang/rust/pull/45991

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by gnzlbg** [comment]
@japaric might be worth to retry this today or tomorrow, the fix was merged yesterday in rustc :)

**Comment by japaric** [comment]
@homunkulus retry

**Comment by homunkulus** [comment]
:hourglass: Trying commit 81f03cd510640c341df0005cc7f21b86cadae43a with merge 6a7bee866502c7ebeb9533155c6f57ff352c70e3...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/cross/builds/303978506?utm_source=github_status&utm_medium=notification)

**Comment by japaric** [comment]
linker error:

> /checkout/src/liballoc_jemalloc/../jemalloc/src/jemalloc.c:797: undefined reference to `secure_getenv'

that sounds like the std facade is being built against a newer libc than the one we are using. The easiest fix would be to bump our Ubuntu version to match the one that's being used in rust-lang/rust. The fix that would make i586 binaries more compatible would be to lower the libc version used in rust-lang/rust.

**Comment by gnzlbg** [comment]
I've bumped the ubuntu version to 16.04 (the one used in rust-lang/rust). If that works we could try lowering it to 14.04 and see if that works.

**Comment by japaric** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 7f3bfb6b42f2420806e6ba451abaa1e86f065fb5 with merge 04e7fecc00f8d3784ebcab5b051643e48247a2b1...

**Comment by japaric** [comment]
@gnzlbg you can `try` this PR, and any other PR, by the way. It didn't work before because the branch was in a `try (failed)` stage. In that stage you have to `retry`. I think you should be able to retry `try`-ed PRs but I've never tested that before.

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/cross/builds/303993985?utm_source=github_status&utm_medium=notification)
State: approved= try=True

**Comment by japaric** [comment]
We shouldn't use a libc older than the one used by rust-lang/rust because there's the risk that a new libc symbol will be used in the std facade in the future and that will cause linker error due to us having an older libc without that symbol.

If you uncomment the other build jobs I'll r+ this.

**Comment by gnzlbg** [comment]
@japaric I've rebased but I think the versions of libc in the readme are still wrong.

**Comment by gnzlbg** [comment]
@japaric so i've updated the version of glibc in the readme to 2.23 and gcc to 5.3.1 (those seem to be the values for ubuntu 16.04). IIUC the openssl version is 1.0.2m for all right?

**Comment by gnzlbg** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit ae1f3d5a7f094969525fe9a6b53398abda635766 with merge 999cc8945a9a62157c23ae20d9a8fb8d8571b349...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/japaric/cross/builds/304387395?utm_source=github_status&utm_medium=notification)
State: approved= try=True

**Comment by japaric** [comment]
>  IIUC the openssl version is 1.0.2m for all right?

Yes. The OpenSSL is the same on all the images.

Thank you, @gnzlbg.

**Comment by gnzlbg** [comment]
@japaric could you do a release of this? I'd like to move `stdsimd` to cross asap, even if that means loosing `arm-unknown-linux-gnueabihf` support temporarily for the time being.

**Comment by japaric** [comment]
@gnzlbg  v0.1.14 has been released. You'll have to wait like 1 hour until propely tagged images make it to dockerhub and the release works properly. When [this build](https://travis-ci.org/japaric/cross/builds/305538558) is done the new release should work.

**Comment by gnzlbg** [comment]
Thanks, I'll send a PR to `stdsimd` to start using `cross` probably tomorrow. I'll ping you in there.

**Pull request merged**